File: packages/control-contract/src/wake.ts
import type { TypedId } from "./ids.js";
export type WakeSource = "assignment" | "comment" | "mention" | "routine" | "approval_result" | "retry" | "recovery" | "manual";
export type WakeRequest = { readonly id: string; readonly work_item_id: TypedId; readonly sources: readonly WakeSource[]; readonly requested_at: string; readonly claimed_at: string | null };
/** Lease ownership is distinct from executor run binding and expires independently. */
export type WorkCheckoutLease = { readonly work_item_id: TypedId; readonly owner_agent_instance_id: TypedId; readonly attempt_id: TypedId; readonly expires_at: string };

File: src-tauri/crates/altai-control-protocol/src/wake.rs
//! Durable wake queue and exclusive checkout lease contracts.

use crate::{AgentInstanceId, AttemptId, WorkItemId};
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum WakeSource {
    Assignment,
    Comment,
    Mention,
    Routine,
    ApprovalResult,
    Retry,
    Recovery,
    Manual,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct WakeRequest {
    pub id: String,
    pub work_item_id: WorkItemId,
    pub sources: Vec<WakeSource>,
    pub requested_at: String,
    pub claimed_at: Option<String>,
}

/// Exclusive work ownership. This is not a RunBinding: a lease can expire or
/// be compare-and-cleared independently of the executor's run lifecycle.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct WorkCheckoutLease {
    pub work_item_id: WorkItemId,
    pub owner_agent_instance_id: AgentInstanceId,
    pub attempt_id: AttemptId,
    pub expires_at: String,
}
